java/driver/flight-sql/src/main/java/org/apache/arrow/adbc/driver/flightsql/ObjectMetadataBuilder.java

```diff
@@ -0,0 +1,239 @@
+import org.apache.arrow.vector.ipc.ReadChannel;
+import org.apache.arrow.vector.ipc.message.MessageSerializer;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+final class ObjectMetadataBuilder {
+
+  private final FlightSqlClient client;
+  private final VectorSchemaRoot root;
+  private final VarCharVector adbcCatalogNames;
+  private final UnionListWriter adbcCatalogDbSchemasWriter;
+  private final BaseWriter.StructWriter adbcCatalogDbSchemasStructWriter;
+  private final BaseWriter.ListWriter adbcCatalogDbSchemaTablesWriter;
+  private final VarCharWriter adbcCatalogDbSchemaNameWriter;
+  private final BaseWriter.StructWriter adbcTablesStructWriter;
+  private final VarCharWriter adbcTableNameWriter;
+  private final VarCharWriter adbcTableTypeWriter;
+  private final BaseWriter.ListWriter adbcTableColumnsWriter;
+  private final BufferAllocator allocator;
+  private final AdbcConnection.GetObjectsDepth depth;
+  private final String catalogPattern;
+  private final String dbSchemaPattern;
+  private final String tableNamePattern;
+  private final String[] tableTypes;
+  private final String columnNamePattern;
+
+  ObjectMetadataBuilder(
+      BufferAllocator allocator,
+      FlightSqlClient client,
+      final AdbcConnection.GetObjectsDepth depth,
+      final String catalogPattern,
+      final String dbSchemaPattern,
+      final String tableNamePattern,
+      final String[] tableTypes,
+      final String columnNamePattern) {
+    this.allocator = allocator;
+    this.client = client;
+    this.depth = depth;
+    this.catalogPattern = catalogPattern;
+    this.dbSchemaPattern = dbSchemaPattern;
+    this.tableNamePattern = tableNamePattern;
+    this.tableTypes = tableTypes;
+    this.columnNamePattern = columnNamePattern;
+    this.root = VectorSchemaRoot.create(StandardSchemas.GET_OBJECTS_SCHEMA, allocator);
+    this.adbcCatalogNames = (VarCharVector) root.getVector(0);
+    this.adbcCatalogDbSchemasWriter = ((ListVector) root.getVector(1)).getWriter();
+    this.adbcCatalogDbSchemasStructWriter = adbcCatalogDbSchemasWriter.struct();
+    this.adbcCatalogDbSchemaTablesWriter =
+        adbcCatalogDbSchemasStructWriter.list("db_schema_tables");
+    this.adbcCatalogDbSchemaNameWriter = adbcCatalogDbSchemasStructWriter.varChar("db_schema_name");
+    this.adbcTablesStructWriter = adbcCatalogDbSchemaTablesWriter.struct();
+    this.adbcTableNameWriter = adbcTablesStructWriter.varChar("table_name");
+    this.adbcTableTypeWriter = adbcTablesStructWriter.varChar("table_type");
+    this.adbcTableColumnsWriter = adbcTablesStructWriter.list("table_columns");
+  }
+
+  private void writeVarChar(VarCharWriter writer, String value) {
+    byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
+    try (ArrowBuf tempBuf = allocator.buffer(bytes.length)) {
+      tempBuf.setBytes(0, bytes, 0, bytes.length);
+      writer.writeVarChar(0, bytes.length, tempBuf);
+    }
+  }
```

Comment:
I wonder why we don't just upstream this (at some point). This bit of code gets written so much for no gain...

Reply:
I'll take a stab at it later. `writeVarChar(Text text)` method is actually part of `VarCharWriterImpl` already, but it's not exposed through `VarCharWriter` interface for some reason. Most of the codebase around writers is code generated, so this might actually be an oversight.

Reply:
@jduo looks like you already fixed this with [38614](https://github.com/apache/arrow/issues/38614), but adbc still depends on version 14. Can we bump arrow dependency version as part of this PR or should that be separate?